core/blockchain/impl/block_tree_impl.cpp

```diff
@@ -72,6 +72,33 @@ namespace kagome::blockchain {
     return !(*this == other);
   }
 
+  BlockTreeImpl::TreeMeta::TreeMeta(TreeNode &subtree_root_node)
+      : deepest_leaf{subtree_root_node}, last_finalized{subtree_root_node} {
+    std::function<void(std::shared_ptr<TreeNode>)> handle =
+        [&](std::shared_ptr<TreeNode> node) {
+          // avoid of deep recurse
+          while (node->children.size() == 1) {
+            node = node->children.front();
+          }
+
+          // is leaf
+          if (node->children.empty()) {
+            leaves.emplace(node->block_hash);
+
+            if (node->depth > deepest_leaf.get().depth) {
+              deepest_leaf = *node;
+            }
+          } else {
+            // follow descendants recursively
+            for (const auto &child : node->children) {
+              handle(child);
```

Comment:
I am somewhat worried whether the block tree may become big enough to still produce a recursion that is too deep for the stack limitations. However if it happens, we'll be able to fix it easily, so I don't mind keeping it the way it is.

Reply:
I don't think we might get that deep block tree, but yes good to keep that in mind